spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,483 @@
+from snorkel.classification.scorer import Scorer
+
+spam_task = Task(
+    name="spam_task",
+    module_pool=module_pool,
+    task_flow=task_flow,
+    loss_func=partial(ce_loss, "head_op"),
+    output_func=partial(softmax, "head_op"),
+    scorer=Scorer(metrics=["accuracy", "f1"]),
+)
+
+# %% [markdown]
+# We'll initialize a [`SnorkelClassifier`](https://snorkel.readthedocs.io/en/redux/source/snorkel.classification.html) with the `spam_task` we've created, initialize a corresponding [`Trainer`](https://snorkel.readthedocs.io/en/redux/source/snorkel.classification.training.html#module-snorkel.classification.training.trainer), and `fit` to our dataloaders!
+
+# %%
+from snorkel.classification.snorkel_classifier import SnorkelClassifier
+from snorkel.classification.training import Trainer
+
+model = SnorkelClassifier([spam_task])
+trainer = Trainer(n_epochs=5, lr=1e-4, progress_bar=True)
+# trainer.fit(model, [dl_train, dl_valid])
+
+# %% [markdown]
+# How well does our model do?
+
+# %%
+model.score([dl_train, dl_valid], as_dataframe=True)
+
+# %% [markdown]
+# ## 2. Perform error analysis
+#
+# In overall metrics (`f1`, `accuracy`) our model appears to perform well!
+#
+# However, we emphasize here that more often than not, we're interested in performance for application-critical subsets, or _slices_.
+#
+# Let's perform an [`error_analysis`](https://snorkel.readthedocs.io/en/redux/source/snorkel.analysis.html#module-snorkel.analysis.error_analysis) to see where our model makes mistakes.
+# We'll collect the predictions from the model and visualize examples in specific error buckets.
+
+# %%
+from snorkel.analysis.error_analysis import get_label_buckets
+from snorkel.analysis.utils import probs_to_preds
+
+outputs = model.predict(dl_valid, return_preds=True)
+error_buckets = get_label_buckets(
+    outputs["golds"]["spam_task"], outputs["preds"]["spam_task"]
+)
+
+# %% [markdown]
+# For application purposes, we might care especially about false negatives (true label was `1`, but model predicted `0`) — for the spam task, external links might point to malware, and we don't want to expose our users to these risks!
+
+# %%
+df_valid[["text", "label"]].iloc[error_buckets[(1, 0)]].head()
+
+# %% [markdown]
+# We notice that we're mis-classifying particularly some comments with shortened urls (e.g. `bit.ly/...`) — these links could redirect us to potentially dangerous websites, and we don't want our users to click them!
+
+# %% [markdown]
+# ## 3. Monitor data slices
+#
+# We leverage *slicing functions* (SFs) — an abstraction that shares syntax with *labeling functions*, which you should already be familiar with! (If not, please see the [intro tutorial](https://github.com/snorkel-team/snorkel-tutorials/blob/master/spam/01_spam_tutorial.ipynb).) A key difference: whereas labeling functions output labels, slicing functions output binary _masks_ indicating whether an example is in the slice or not.
```

Comment:
slicing functions actually output binary indicators, right? then the collection of those makes a mask.